src/test/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounderTest.java

```diff
@@ -97,14 +112,228 @@ public void groundConstraintAlreadyGround() {
+		ThriceTruth[] truthsOfB = new ThriceTruth[nTrueBs];
+		for (int i = 0; i < nTrueBs; i++) {
+			truthsOfB[i] = ThriceTruth.TRUE;
+		}
+		testLaxGrounderHeuristicTolerance(program, ruleID, startingLiteral, startingInstance, tolerance, truthsOfB, 1, expectNoGoods, expectedNumbersOfUnassignedPositiveBodyAtoms);
+	}
+
+	private void testLaxGrounderHeuristicTolerance(Program program, int ruleID, Literal startingLiteral, int startingInstance, int tolerance, ThriceTruth[] truthsOfB, int arityOfB, boolean expectNoGoods, List<Integer> expectedNumbersOfUnassignedPositiveBodyAtoms) {
+		AtomStore atomStore = new AtomStoreImpl();
+		TrailAssignment currentAssignment = new TrailAssignment(atomStore);
+		GrounderHeuristicsConfiguration heuristicConfiguration = GrounderHeuristicsConfiguration.getInstance(tolerance, tolerance);
+		NaiveGrounder grounder = (NaiveGrounder) GrounderFactory.getInstance("naive", program, atomStore, p -> true, heuristicConfiguration, true);
+
+		int[] bAtomIDs = new int[truthsOfB.length];
+		for (int i = 0; i < truthsOfB.length; i++) {
+			int[] bTerms = new int[arityOfB];
+			for (int n = 0; n < arityOfB; n++) {
+				bTerms[n] = (n == arityOfB - 1) ? i + 1 : startingInstance;
+			}
+			bAtomIDs[i] = atomStore.putIfAbsent(atom("b", bTerms));
+		}
+		addAtomsToWorkingMemoryWithoutChangingTheAssignment(atomStore, grounder, bAtomIDs);
+		assign(currentAssignment, bAtomIDs, truthsOfB);
+
+		grounder.bootstrap();
+		final NonGroundRule nonGroundRule = grounder.getNonGroundRule(ruleID);
+		final Substitution substStartingLiteral = Substitution.unify(startingLiteral, new Instance(ConstantTerm.getInstance(startingInstance)), new Substitution());
+		final NaiveGrounder.BindingResult bindingResult = grounder.getGroundInstantiations(nonGroundRule, nonGroundRule.groundingOrder.groundingOrders.get(startingLiteral), substStartingLiteral, currentAssignment);
+		assertEquals(expectNoGoods, bindingResult.size() > 0);
+		if (bindingResult.size() > 0) {
+			assertEquals(expectedNumbersOfUnassignedPositiveBodyAtoms, bindingResult.numbersOfUnassignedPositiveBodyAtoms);
+		} else {
+			assertTrue(bindingResult.numbersOfUnassignedPositiveBodyAtoms.isEmpty());
+		}
+	}
+
+	private void assign(TrailAssignment currentAssignment, int[] atomIDs, ThriceTruth[] truthValues) {
+		currentAssignment.growForMaxAtomId();
+		for (int i = 0; i < truthValues.length; i++) {
+			int atomID = atomIDs[i];
+			if (truthValues[i] != null) {
+				currentAssignment.assign(atomID, truthValues[i]);
+			}
+		}
+	}
+
+	private void addAtomsToWorkingMemoryWithoutChangingTheAssignment(AtomStore atomStore, NaiveGrounder grounder, int[] atomIDs) {
+		TrailAssignment temporaryAssignment = new TrailAssignment(atomStore);
+		temporaryAssignment.growForMaxAtomId();
+		for (int b : atomIDs) {
+			temporaryAssignment.assign(b, TRUE);
+		}
+		grounder.updateAssignment(temporaryAssignment.getNewPositiveAssignmentsIterator());
+	}
```

Comment:
Those methods are called only once and each iterates over an array of integers constructed in lines +295 to +304. Why not inline those methods and remove the double array-creation/iteration?

Reply:
I am not sure what you mean by "inline" and where you see a "double array-creation", but those two methods are concerned with different things and need to be executed one after the other to produce the desired result.